### PR TITLE
Tor

### PR DESCRIPTION
**Pull request by nochiel**
This PR depends on <https://github.com/breez/breezmobile/pull/546>.

- [x] Add Tor for LND.
    - [x] Update dependency to [btcd v0.15.0 ](https://github.com/lightningnetwork/lnd/projects/14#card-77040604)so that we can use Tor v3.
- [x] WebDAV backups use Tor.
    - [x] Test using <http://files.com>. 
    - [x] Test using <http://nextcloud.com>. 
- [x] Test: Verify that user can connect to their own Bitcoin node over Tor.
  - [x] Connect to Bitcoin node using onion address.
    - Run `Peers> connect` from the Developers screen to connect to your own node. 
        - This will work for Tor Hidden Service nodes.
- [x] Enable/Disable Tor.
    - [x] Ask user to restart Breez with new configuration.
- [ ] LNURLs can be onion addresses and should go through Tor.
    - [x] LNUrlAuth
    - [x] LNUrlWithdraw
    - [x] LNUrlPay
- Should all other http requests go through Tor?
    - [ ] Submit PR to fiatjaf/go-lnurl to allow users to configure the http client that it uses.

## Fix
- [x] Make changes requested by @yaslama in code review.
    - [ ] [Use enums for get/setTorStatus](https://github.com/breez/breez/pull/139#discussion_r687706400)
- [x] Bug: Payments time out over Tor. Why?
    - I think this is because I was attempting payments before LND had finished syncing.
    - [x] Test: When running Tor, increase the timeout so that payments can be completed.
- [x] Bug: BreezMobile will report that toggling Tor has failed even when it succeeds.
- ~[ ] Disabling Tor doesn't always manage to restart the daemon.~
- [x] Bug: Enabling Tor and restarting Breezmobile doesn't enable Tor for Webdav.
  - ~To refresh the backup provider, user has to change backup setting from `remote server` to `google drive` and back again. smh.~
  - [x] Fix: reset the current backup provider's Tor configuration.
- Error handling:
    - [x] If Tor is enabled but fails to start, how should we show the user an error?

## Test
- Sync over Tor is very slow.
    - [x] Test that BreezMobile can actually complete syncing over Tor.
- [x] Verify opening channels and payments work consistently/correctly.
  - Ref. [Opening channels to a node over Tor](https://github.com/breez/breezmobile/wiki/Opening-Channels)

## iOS
- ~[Implement iOS wrapper for upstream go-libtor dependency](https://github.com/ipsn/go-libtor/issues/3)~
    - ~Please could a MacOS user check if this is necessary or does go-libtor compile out of the box for iOS?~
    - ~using `berty.tech/go-libtor` bindings which have iOS/Darwin support.~
- [x] Integrate Tor in iOS using [Tor Framework](https://github.com/iCepa/Tor.framework).

### CODE REVIEW

**Comment by nochiel** [comment]
@roeierez Please review and tell me if the current implementation makes sense. What functionality should I add/change?

File: account/connect.go [hunk 1]
```diff
@@ -235,7 +235,7 @@ func (a *Service) openChannel(l lsp, force bool) error {
 	a.log.Info("openChannel started...")
 	_, err, _ := createChannelGroup.Do("createChannel", func() (interface{}, error) {
 		err := a.connectAndOpenChannel(l, force)
-		a.log.Info("openChannel finished, error = %v", err)
+		a.log.Infof("openChannel finished, error = %v", err)
```
Comment (yaslama) [inline account/connect.go:238]: Can you please put all the small fixes in a separate commit? If you can put them in a separate PR, that would be even better.

File: app.go [hunk 2]
```diff
@@ -397,3 +433,18 @@ func (a *App) CheckLSPClosedChannelMismatch(
 	}
 	return &data.CheckLSPClosedChannelMismatchResponse{Mismatch: mismatch}, nil
 }
+
+func (a *App) EnableOrDisableTor(enable bool) error {
```
Comment (yaslama) [inline app.go:437]: The names EnableOrDisableTor and IsTorActive are not consistent. I suggest to use in all places SetTor() and GetTor() where the SetTor parameter and the return value of GetTor is an enum constant from the breez/tor package (use the first iota as undefined because it's the default value).

File: data/messages.pb.go [hunk 3]
```diff
@@ -1,17 +1,12 @@
 // Code generated by protoc-gen-go. DO NOT EDIT.
 // versions:
-// 	protoc-gen-go v1.25.0
-// 	protoc        v3.15.8
+// 	protoc-gen-go v1.26.0
```
Comment (yaslama) [inline data/messages.pb.go:3]: Why these changes are needed?

File: tor/tor.go [hunk 4]
```diff
@@ -0,0 +1,190 @@
+// TODO(nochiel) Add useful commands to manage Tor to the developer panel.
```
Comment (yaslama) [inline tor/tor.go:1]: This package can be added in a separate commit

**Comment by yaslama** [comment]
@nochiel The PR is very good. I wrote some comments in the diff itself.

**Comment by nochiel** [comment]
> @nochiel The PR is very good. I wrote some comments in the diff itself.

@yaslama Thank you for your feedback! I'll work on the things you mentioned.

File: data/messages.pb.go (hunk 3 above)
Comment (nochiel) [reply data/messages.pb.go]: They aren't. But I couldn't get previous versions of protoc to generate files on my machine. These changes just indicate the versions used to generate the files. But they aren't strictly needed for anything other than documentation, I believe.

File: app.go (hunk 2 above)
Comment (nochiel) [reply app.go:445]: If it's okay with you, can I do this *after* all code review is complete and someone other than myself has tested tor satisfactorily and @kingonly has verified that it's feature complete?

(Right now, whenever I change `api.go` my build becomes fragile and breaks. I'm not yet sure if I need to reinstall Android Studio or roll back an upgrade or if something else is the matter. But it makes me not want to change the interface for now while everything is still working correctly.)

**Comment by openoms** [comment]
Looking forward to this feature!

Found out that LND is not able to connect to Tor v3 neutrino servers as in v0.13.1: https://github.com/lightningnetwork/lnd/issues/5788

Will Breez support Tor v3 backends with this PR?

**Comment by Mohsen7s** [comment]
The only reason Ill switch to use this app its the Tor feature.
Drop tor support completely or support v3; because tor without v3 support its just useless. 
Any update or milestone on this issue ?

**Comment by kingonly** [comment]
I think you can see this isn't merge because we're waiting for Neutrino to support Tor v3: https://github.com/lightningnetwork/lnd/issues/4973

File: backup/drive.go [hunk 5]
```diff
@@ -439,6 +443,14 @@ func (p *GoogleDriveProvider) deleteStaleSnapshots(nodeFolderID, activeBackupFol
 	return nil
 }
 
+func (p *GoogleDriveProvider) SetTor(torConfig *tor.TorConfig) {
+	return
```
[the remark this answers is not in the document]
Comment (nochiel) [reply backup/drive.go:447]: `SetTor` is part of the `Provider` interface. For Google Drive backup provider, it's a no-op.

File: backup/manager_test.go [hunk 6]
```diff
@@ -63,13 +65,21 @@ func (m *MockTester) DownloadBackupFiles(nodeID, backupID string) ([]string, err
 	return m.DownloadBackupFilesImpl(nodeID, backupID)
 }
 
+func (m *MockTester) SetTor(torConfig *tor.TorConfig) {
+	m.torConfig = torConfig
+}
+
+func (m *MockTester) TestAuth() {
+	return
```
[the remark this answers is not in the document]
Comment (nochiel) [reply backup/manager_test.go:73]: I guess one could add a test but it's unclear how useful it would be. That is, the WebDav provider requires us to test authentication but other providers don't.

File: backup/drive.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (ubbabeck) [reply backup/drive.go:447]: I see thank you!

File: bindings/native_backup_provider.go [hunk 7]
```diff
@@ -72,9 +73,17 @@ func (b *NativeBackupProviderBridge) DownloadBackupFiles(nodeID, backupID string
 	return strings.Split(files, ","), nil
 }
 
+func (b *NativeBackupProviderBridge) SetTor(torConfig *tor.TorConfig) {
+	return
```
[the remark this answers is not in the document]
Comment (nochiel) [reply bindings/native_backup_provider.go:77]: My understanding is that iOS uses the NativeBackupProvider to back up to iCloud. Is there any reason to use Tor when accessing iCloud, assuming that accessing a user's iCloud account over Tor will even work consistently?

[the remark this answers is not in the document]
Comment (nochiel) [reply bindings/native_backup_provider.go:77]: > We shouldn't use Tor for iCloud, just WebDAV is relevant and only in case the server is using Tor.

That makes sense. WebDAV already works over Tor.

**Comment by nochiel** [comment]
@kingonly @ubbabeck  Shouldn't this PR be merged now that the LND dependency has been updated in master and tor v3 support is available?
